solidity/contracts/CosmosToken.sol

```diff
@@ -46,14 +47,15 @@ contract CosmosERC20 is ERC20 {
 
 
 	/**
-	 * @dev Sets the gravity contract to a new address.
+	 * @dev Sets the gravity contract and transfer the balance to the new address.
 	 *
 	 * Requirements:
 	 *
 	 * - `msg.sender` must be the current gravity contract
 	 */
 	function setGravityContract(address _gravityAddress) external onlyGravity {
-
+		require(balanceOf(_gravityAddress) == 0, "new gravity address balance should be zero");
```

Comment:
Looks good to me, apart from this, think adding some other checks like below could be better:
- repeat check: `require(_gravityAddress != msg.sender);`
- non-zero address check: `require(_gravityAddress != address(0));`
- this address check: `require(_gravityAddress != address(this));`
- something like owner check: `require(_gravityAddress != IGravity(msg.sender).owner());` (optional for you: need to add corresponding interface of Gravity.sol)

Reply:
- for repeat check : is it redundant with `balanceOf(_gravityAddress) == 0?`
- for non-zero address check, I think it should be done in `_transfer`
- `this address check` may be harmless?

Reply:
based on the order of you message:
- I just mean contract address repeat(by a mistaken), but this address actually can be any address you want(e.g. normal wallet). Looks ok to me, it's an optional one too, up to your decision ^_^.
- oh, agree. I forgot the basic one in erc20. Can Ignore.
- just aiming to avoid the mistake of passing address of this. Looks ok, up to your decision, can ignore that ^_^.

Reply:
I'm looking if there is a way to enforce that the _gravityAddress is a contract type `IGravity`

Seems could use https://eips.ethereum.org/EIPS/eip-165 but its quite a change

Reply:
Please note that Calvin & Landan's concern about the existing VVStoken(old version of CosmosToken).
It means that the function interface format is fixed as `function setGravityContract(address _gravityAddress) external`. 
If you force that the _gravityAddress is a contract type `IGravity` like `function setGravityContract(IGravity _gravityAddress) external`, you have to separate the workflow in Gravity.sol for corresponding to the two kinds of CosmosToken and plus the normal ERC20
<img width="1349" alt="Screen Shot 2022-08-04 at 17 59 23" src="https://user-images.githubusercontent.com/86825449/182820220-f0bb5e72-782e-430b-b93d-132993ca5f26.png">
.